invokeai/app/services/model_records/model_records_sql.py

```diff
@@ -0,0 +1,396 @@
+# Copyright (c) 2023 Lincoln D. Stein and the InvokeAI Development Team
+"""
+SQL Implementation of the ModelRecordServiceBase API
+
+Typical usage:
+
+  from invokeai.backend.model_manager import ModelConfigStoreSQL
+  store = ModelConfigStoreSQL(sqlite_db)
+  config = dict(
+        path='/tmp/pokemon.bin',
+        name='old name',
+        base_model='sd-1',
+        type='embedding',
+        format='embedding_file',
+     )
+
+   # adding - the key becomes the model's "key" field
+   store.add_model('key1', config)
+
+   # updating
+   config.name='new name'
+   store.update_model('key1', config)
+
+   # checking for existence
+   if store.exists('key1'):
+      print("yes")
+
+   # fetching config
+   new_config = store.get_model('key1')
+   print(new_config.name, new_config.base_model)
+   assert new_config.key == 'key1'
+
+  # deleting
+  store.del_model('key1')
+
+  # searching
+  configs = store.search_by_path(path='/tmp/pokemon.bin')
+  configs = store.search_by_hash('750a499f35e43b7e1b4d15c207aa2f01')
+  configs = store.search_by_name(base_model='sd-2', model_type='main')
+"""
+
+
+import json
+import sqlite3
+import threading
+from pathlib import Path
+from typing import List, Optional, Union
+
+from invokeai.backend.model_manager.config import (
+    AnyModelConfig,
+    BaseModelType,
+    ModelConfigBase,
+    ModelConfigFactory,
+    ModelType,
+)
+
+from ..shared.sqlite import SqliteDatabase
+from .model_records_base import (
+    CONFIG_FILE_VERSION,
+    DuplicateModelException,
+    ModelRecordServiceBase,
+    UnknownModelException,
+)
+
+
+class ModelRecordServiceSQL(ModelRecordServiceBase):
+    """Implementation of the ModelConfigStore ABC using a SQL database."""
+
+    _conn: sqlite3.Connection
+    _cursor: sqlite3.Cursor
+    _lock: threading.Lock
+
+    def __init__(self, db: SqliteDatabase):
+        """
+        Initialize a new object from preexisting sqlite3 connection and threading lock objects.
+
+        :param conn: sqlite3 connection object
+        :param lock: threading Lock object
+        """
+        super().__init__()
+        self._conn = db.conn
+        self._lock = db.lock
+        self._conn.row_factory = sqlite3.Row
+        self._cursor = self._conn.cursor()
```

Comment:
Why split the `SqliteDatabase` up into `self._conn`, `self._lock`, and `self._cursor`? I'd just keep it all together as `self._db`.

Reply:
Just following the pattern in `image_records_sqlite.py` and so forth.

I've revised to keep the lock and connection in the db object. However, I continue to break out cursor() because this initializes a new cursor object that we want to reuse.

Reply:
Interesting. I just took a closer look at how we are accessing the DB broadly across the app. My interpretation of the sqlite3 docs is that cursors are cheap to create and a new cursor should be used for each 'interaction' - though the docs aren't 100% clear on this point. Now I'm wondering if this sharing of a cursor is the reason that we need the DB thread lock in the first place.

No need to address in this PR, as we aren't making it any worse than it already was. I'll try to find some time to look into this some more.

Reply:
Right. Closing conversation.